### PR TITLE
Support running on Node.js

### PR DESCRIPTION
**Pull request by nex3**
See sass/migrator#56

### CODE REVIEW

**Comment by nex3** [comment]
Can you publish @srawlins?

**Comment by srawlins** [comment]
I tagged it in GitHub but...

```
Looks great! Are you ready to upload your package (y/n)? y
Uploading... (2.0s)
InsufficientPermissions: Insufficient permissions to upload new versions of package `glob`..
```

☹️

**Comment by natebosch** [comment]
Please hold off on publishing. I'm not sure that our build systems support exporting `'dart:io'`. Also we need to consider whether we want this extra dep internally, and in flutter etc.

**Comment by jakemac53** [comment]
Yes `build_runner` would require a layer of indirection here instead of directly exporting `dart:io`.

**Comment by natebosch** [comment]
Another concern here, assuming we could support the conditional export, this will make the import go from a build time failure on the web, to a runtime failure. If my understanding is correct this will only work on node, but the condition also matches the web.

File: lib/src/io.dart [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// These libraries don't expose *exactly* the same API, but they overlap in all
+// the cases we care about.
+export 'dart:io' if (dart.library.js) 'package:node_io/node_io.dart';
```
Comment (jakemac53) [inline lib/src/io.dart]: This will have the effect of selecting `package:node_io/node_io.dart` for all _web_ projects too, without any analysis errors...

Comment (jakemac53) [reply lib/src/io.dart]: nate beat me to it, resolving

**Comment by nex3** [comment]
What do we need to get this published quickly? Glob support is important for the Sass module migrator, which is launching next week. I'm definitely open to alternate ways of injecting `node_io`, if you have other suggestions.

> Another concern here, assuming we could support the conditional export, this will make the import go from a build time failure on the web, to a runtime failure. If my understanding is correct this will only work on node, but the condition also matches the web.

I thought that `dart:io` was already supported by dart2js at build-time, and just threw runtime errors for all API usage. If there's a better condition to use in the import, though, we can definitely change it.

**Comment by jakemac53** [comment]
Long term this seems like a really good use case for extension methods - we could move all the apis that require file access to extension methods so the normal `Glob` class could be available for web and actually work.

The extensions would exist in custom entrypoints `glob_node.dart`, `glob_io.dart` etc. Or possibly we could conditionally export the extension from the main entrypoint (I am not 100% sure that would work, but it probably would).

> I thought that `dart:io` was already supported by dart2js at build-time, and just thre runtime errors for all API usage. If there's a better condition to use in the import, though, we can definitely change it.

`build_web_compilers` specifically won't allow it - even though technically the compilers themselves still do. We migrated all the original use cases that needed this to use conditional imports.

> What do we need to get this published quickly? Glob support is important for the Sass module migrator, which is launching next week. I'm definitely open to alternate ways of injecting `node_io`, if you have other suggestions.

Presumably `dart:html` doesn't exist on node? So we could have an extra condition that exports a stub which throws so at least we don't try and compile `node_io` for web. This won't give us any build time failures still, but it would be slightly better.

In general it is also advised to extract the logic you need from `dart:io` into top level functions which you implement in all conditional imports, as opposed to trying to conditionally export libraries that do not expose the same api like this. That would make stubbing easier as well.

**Comment by nex3** [comment]
> Presumably `dart:html` doesn't exist on node? So we could have an extra condition that exports a stub which throws so at least we don't try and compile `node_io` for web. This won't give us any build time failures still, but it would be slightly better.

If you export `dart:io` when in `dart:html` mode, will you get a build-time failure when compiling for the web?

> In general it is also advised to extract the logic you need from `dart:io` into top level functions which you implement in all conditional imports, as opposed to trying to conditionally export libraries that do not expose the same api like this. That would make stubbing easier as well.

This makes sense in general, but in this case `node_io` is a subset of `dart:io`'s API that's explicitly intended to be used as a drop-in replacement.

**Comment by natebosch** [comment]
> What do we need to get this published quickly?

At a minimum we need to see some examples of how this behaves both internally and with `build_runner`. I don't know if we've ever tried `export 'dart:io';` and I don't know what could be broken or how.

> If you export dart:io when in dart:html mode, will you get a build-time failure when compiling for the web?

I think so? This feels really strange but could be better than nothing.

**Comment by jakemac53** [comment]
> > If you export dart:io when in dart:html mode, will you get a build-time failure when compiling for the web?

> I think so? This feels really strange but could be better than nothing.

I believe we would eagerly skip even trying to compile because of the dart:io dependency.

**Comment by nex3** [comment]
> I believe we would eagerly skip even trying to compile because of the dart:io dependency, and we would log a warning about it.

That's the behavior we want, right?

**Comment by jakemac53** [comment]
> That's the behavior we want, right?

Correct - note that the export needs to be hidden behind a layer of indirection for build_runner to support it today (we don't allow sdk imports/exports in conditionals). I can't remember exactly why we have that restriction.

**Comment by nex3** [comment]
So if we put `dart:io` behind a file that just contains `export "dart:io"` and added explicit handling for `dart.library.html`, would this be safe to release?

**Comment by jakemac53** [comment]
Yes I think that is pretty much the best we can do right now. I verified locally that it works as expected and will send it out as a PR.

**Pull request merged**